lib/Arch/AArch32/Semantics/BINARY.cpp

```diff
@@ -0,0 +1,540 @@
+  auto rhs = ZExt(Read(src3));
+  auto lhs = ZExt(Read(src2));
+  auto acc_hi = ZExt(Read(src1));
+  auto acc_lo = ZExt(Read(src4));
+  auto res = UAdd(UAdd(UMul(lhs, rhs), acc_hi), acc_lo);
+  Write(dst_lo, TruncTo<uint32_t>(res));
+  Write(dst_hi, TruncTo<uint32_t>(UShr(res, 32ul)));
+  return memory;
+}
+
+DEF_COND_SEM(MLS, R32W dst, R32 src1, R32 src2, R32 src3) {
+  auto rhs = Signed(Read(src2));
+  auto lhs = Signed(Read(src1));
+  auto acc = Signed(Read(src3));
+  auto res = Unsigned(SSub(acc, SMul(lhs, rhs)));
+  Write(dst, res);
+  return memory;
+}
+
+DEF_COND_SEM(UMULL, R32W dst_hi, R32W dst_lo, R32 src1, R32 src2, R32 src3, R32 src4) {
+  auto rhs = ZExt(Read(src3));
+  auto lhs = ZExt(Read(src2));
+  auto acc = UOr(UShl(ZExt(Read(src1)), 32ul), ZExt(Read(src4))); // UInt(R[dHi]:R[dLo])
+  auto res = UAdd(UMul(lhs, rhs), acc);
+  Write(dst_hi, TruncTo<uint32_t>(UShr(res, 32ul)));
+  Write(dst_lo, TruncTo<uint32_t>(res));
+  return memory;
+}
+
+DEF_COND_SEM(UMULLS, R32W dst_hi, R32W dst_lo, R32 src1, R32 src2, R32 src3, R32 src4) {
+  auto rhs = ZExt(Read(src3));
+  auto lhs = ZExt(Read(src2));
+  auto acc = UOr(UShl(ZExt(Read(src1)), 32ul), ZExt(Read(src4))); // UInt(R[dHi]:R[dLo])
+  auto res = UAdd(UMul(lhs, rhs), acc);
+  state.sr.n = SignFlag(res);
+  state.sr.z = ZeroFlag(res);
+  // PSTATE.C, PSTATE.V unchanged
+  Write(dst_hi, TruncTo<uint32_t>(UShr(res, 32ul)));
+  Write(dst_lo, TruncTo<uint32_t>(res));
+  return memory;
+}
+
+DEF_COND_SEM(SMULL, R32W dst_hi, R32W dst_lo, R32 src1, R32 src2, R32 src3, R32 src4) {
+  auto rhs = SExt(Signed(Read(src3)));
+  auto lhs = SExt(Signed(Read(src2)));
+  auto acc = SOr(SShl(SExt(Read(src1)), 32ul), Signed(ZExt(Read(src4)))); // UInt(R[dHi]:R[dLo])
+  auto res = SAdd(SMul(lhs, rhs), acc);
+  Write(dst_hi, TruncTo<uint32_t>(SShr(res, 32ul)));
+  Write(dst_lo, TruncTo<uint32_t>(res));
+  return memory;
+}
+
+DEF_COND_SEM(SMULLS, R32W dst_hi, R32W dst_lo, R32 src1, R32 src2, R32 src3, R32 src4) {
+  auto rhs = SExt(Signed(Read(src3)));
+  auto lhs = SExt(Signed(Read(src2)));
+  auto acc = SOr(SShl(SExt(Read(src1)), 32ul), Signed(ZExt(Read(src4)))); // UInt(R[dHi]:R[dLo])
+  auto res = SAdd(SMul(lhs, rhs), acc);
+  state.sr.n = SignFlag(res);
+  state.sr.z = ZeroFlag(res);
+  // PSTATE.C, PSTATE.V unchanged
+  Write(dst_hi, TruncTo<uint32_t>(SShr(res, 32ul)));
+  Write(dst_lo, TruncTo<uint32_t>(res));
+  return memory;
+}
+} // namespace
+
+DEF_ISEL(MUL) = MUL;
+DEF_ISEL(MULS) = MULS;
+DEF_ISEL(MLA) = MUL;
+DEF_ISEL(MLAS) = MULS;
+DEF_ISEL(MLS) = MLS;
+DEF_ISEL(UMAAL) = UMAAL;
+DEF_ISEL(UMULL) = UMULL;
+DEF_ISEL(UMULLS) = UMULLS;
+DEF_ISEL(UMLAL) = UMULL;
+DEF_ISEL(UMLALS) = UMULLS;
+DEF_ISEL(SMULL) = SMULL;
+DEF_ISEL(SMULLS) = SMULLS;
+DEF_ISEL(SMLAL) = SMULL;
+DEF_ISEL(SMLALS) = SMULLS;
+
+// Halfword Multiply and Accumulate
+namespace {
+DEF_COND_SEM(SMLAh, R32W dst, R32 src1, R32 src2, R32 src3) {
+  auto rhs = SExt<uint64_t>(Read(src2));
```

Comment:
Normally I use `SExtTo` when specifying a destination type. In this case, you've specified `uint64_t` (an unsigned type), but the operator will convert to `int64_t` via `WST` here: https://github.com/lifting-bits/remill/blob/master/include/remill/Arch/Runtime/Operators.h#L733

I think by specifying `uint64_t` here, what's happening is a sign extension to an `int128_t`; see this: https://github.com/lifting-bits/remill/blob/master/include/remill/Arch/Runtime/Types.h#L664-L666

Is this the intention?